### PR TITLE
Date to Timestamp cast

### LINKED ISSUE #4644: Comparing a `Timestamp` to a `Date32` fails
**Describe the bug**
Comparing a `Timestamp` to a `Date32` fails.

**To Reproduce**
Add the following test to [`timestamps.slt`](https://github.com/apache/arrow-datafusion/blob/5c558e9f6b3fe29ef0d5f78ec9465852d020d989/datafusion/core/tests/sqllogictests/test_files/timestamps.slt#L3):

```
# Test that we can compare a timestamp to a string casted to a date
query C rowsort
select * from foo where ts > '2000-01-01'::date;
----
2 2000-02-01T00:00:00
3 2000-03-01T00:00:00
```

This fails w/:

```
[timestamps.slt] Running query: "select * from foo where ts > '2000-01-01'::date;"
Error: SqlLogicTest(query failed: DataFusion error: Internal error: The type of Timestamp(Nanosecond, None) > Date32 of binary physical should be same. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

**Expected behavior**
It works.

**Additional context**
Tested on https://github.com/apache/arrow-datafusion/commit/5c558e9f6b3fe29ef0d5f78ec9465852d020d989 .

This is likely due to a like in [`temporal_coercion`](https://github.com/apache/arrow-datafusion/blob/5c558e9f6b3fe29ef0d5f78ec9465852d020d989/datafusion/expr/src/type_coercion/binary.rs#L539-L539) as well as this workaround:

https://github.com/apache/arrow-datafusion/blob/5c558e9f6b3fe29ef0d5f78ec9465852d020d989/datafusion/optimizer/src/type_coercion.rs#L232-L239

Also see #3419.

### LINKED ISSUE #4761: The coerced type for comparison between timestamp with date
I try the comparison timestamp with date in the Spark, and find the coerced type is timestamp.

```
spark-sql> explain extended select now()>cast('2000-01-01' as date);
== Parsed Logical Plan ==
'Project [unresolvedalias(('now() > cast(2000-01-01 as date)), None)]
+- OneRowRelation

== Analyzed Logical Plan ==
(now() > CAST(2000-01-01 AS DATE)): boolean
Project [(now() > cast(cast(2000-01-01 as date) as timestamp)) AS (now() > CAST(2000-01-01 AS DATE))#49]
+- OneRowRelation

== Optimized Logical Plan ==
Project [true AS (now() > CAST(2000-01-01 AS DATE))#49]
+- OneRowRelation

== Physical Plan ==
*(1) Project [true AS (now() > CAST(2000-01-01 AS DATE))#49]
+- *(1) Scan OneRowRelation[]
```

cc @comphead @alamb 

But current implementation of coerced type is date, we can fix this in the follow up PR.

_Originally posted by @liukun4515 in https://github.com/apache/arrow-datafusion/pull/4741#discussion_r1058277985_

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #4761 
Closes #4644 .

# Rationale for this change
Introduce casting between dates and timestamps datatypes. Currently DF just panics on such kind of queries

# What changes are included in this PR?
This PR includes planner changes to support date to timestamp cast. 

# Are these changes tested?
Yes. Tests enclosed

# Are there any user-facing changes?
No

### CODE REVIEW

**Comment by comphead** [comment]
@liukun4515 @crepererum @alamb please check this one.

The downside for now is additional cast created sometimes for timestamp, to convert `Timestamp(_, Some("00:00"))` to `Timestamp(_, None)`. This can be solved in separate PR if you dont mind as it might again require changes in arrow-rs

File: datafusion/core/tests/sql/timestamp.rs [hunk 1]
```diff
@@ -1742,5 +1741,55 @@ async fn test_ts_dt_binary_ops() -> Result<()> {
 
     assert_batches_eq!(expected, &results);
 
+    //test cast path timestamp date using literals
+    let sql = "select '2000-01-01'::timestamp >= '2000-01-01'::date";
```
Comment (alamb) [inline datafusion/core/tests/sql/timestamp.rs:1745]: You might be able to add these tests to sqllogictest if you wanted -- https://github.com/apache/arrow-datafusion/blob/master/datafusion/core/tests/sqllogictests/test_files/timestamps.slt 

Not required but I find the sqllogictest framework much easier to work with

File: datafusion/expr/src/type_coercion/functions.rs [hunk 2]
```diff
@@ -174,8 +174,8 @@ pub fn can_coerce_from(type_into: &DataType, type_from: &DataType) -> bool {
                 | Float64
                 | Decimal128(_, _)
         ),
-        Timestamp(TimeUnit::Nanosecond, None) => {
-            matches!(type_from, Null | Timestamp(_, None))
+        Timestamp(TimeUnit::Nanosecond, _) => {
+            matches!(type_from, Null | Timestamp(_, _) | Date32)
```
Comment (alamb) [inline datafusion/expr/src/type_coercion/functions.rs:178]: 👍  (I have a draft PR in the works to support coercing from `utf8` as well https://github.com/apache/arrow-datafusion/pull/5117)

File: datafusion/core/tests/sql/timestamp.rs [hunk 3]
```diff
@@ -1742,5 +1741,55 @@ async fn test_ts_dt_binary_ops() -> Result<()> {
...
+    let mut res: Option<String> = None;
+    for row in 0..batch.num_rows() {
+        if &array_value_to_string(batch.column(0), row)?
+            == "logical_plan after type_coercion"
+        {
+            res = Some(array_value_to_string(batch.column(1), row)?);
+            break;
+        }
+    }
+    assert_eq!(res, Some("Projection: CAST(now() AS Timestamp(Nanosecond, None)) = CAST(currentdate() AS Timestamp(Nanosecond, None))\n  EmptyRelation".to_string()));
```
Comment (alamb) [inline datafusion/core/tests/sql/timestamp.rs:1792]: 👍

**Comment by alamb** [comment]
> The downside for now is additional cast created sometimes for timestamp, to convert Timestamp(_, Some("00:00")) to Timestamp(_, None). This can be solved in separate PR if you dont mind as it might again require changes in arrow-rs

I think this is correct (and needed) if the timestamps in question have different timezones. It should not really effect runtime performance as I expect both to be evaluated at plan time

File: datafusion/core/tests/sql/timestamp.rs (hunk 1 above)
Comment (comphead) [reply datafusion/core/tests/sql/timestamp.rs:1745]: Thanks @alamb I will create a separate ticket to move all timestamp tests from timestamp.rs to sqllogictest, all that can be moved of course

**Comment by alamb** [comment]
I plan to merge this PR tomorrow unless there are other comments

File: datafusion/expr/src/type_coercion/binary.rs [hunk 4]
```diff
@@ -541,33 +541,30 @@ fn is_time_with_valid_unit(datatype: DataType) -> bool {
...
+        }
+        (Utf8, Time64(unit)) | (Time64(unit), Utf8) => {
+            match is_time_with_valid_unit(Time64(unit.clone())) {
+                false => None,
+                true => Some(Time64(unit.clone())),
+            }
+        }
+        (Timestamp(_, tz), Utf8) | (Utf8, Timestamp(_, tz)) => {
+            Some(Timestamp(TimeUnit::Nanosecond, tz.clone()))
+        }
```
Comment (waitingkuo) [inline datafusion/expr/src/type_coercion/binary.rs:561]: i tried this, but it didn't work
```bash
❯ select '2000-01-01T00:00:00'::timestamp::timestamptz = '2000-01-01T00:00:00';
Internal("The type of Timestamp(Nanosecond, Some(\"+00:00\")) Eq Utf8 of binary physical should be same")
```

File: datafusion/expr/src/type_coercion/binary.rs [hunk 5]
```diff
@@ -541,33 +541,30 @@ fn is_time_with_valid_unit(datatype: DataType) -> bool {
...
+        (Timestamp(_, None), Date32) | (Date32, Timestamp(_, None)) => {
+            Some(Timestamp(TimeUnit::Nanosecond, None))
+        }
+        (Timestamp(_, _tz), Date32) | (Date32, Timestamp(_, _tz)) => {
+            Some(Timestamp(TimeUnit::Nanosecond, None))
+        }
```
Comment (waitingkuo) [inline datafusion/expr/src/type_coercion/binary.rs:567]: does it mean that we convert Timestamptz to Timestamp and then compare?

**Review by waitingkuo — COMMENTED** [review]
thank you @comphead 
left the comments bellow

I think we could consider adding this coercion as well as it works in postgrseql (perhaps in another tickets)
`(Utf8, Date)`

postgrseql
```bash
willy=# select '2000-01-01T00:00:00' = '2000-01-01'::date;
 ?column? 
----------
 t
(1 row)
```

datafusion
```bash
❯ select '2000-01-01T00:00:00' = '2000-01-01'::date;
ArrowError(CastError("Cannot cast string '2000-01-01T00:00:00' to value of Date32 type"))
```

**Comment by alamb** [comment]
> I think we could consider adding this coercion as well as it works in postgrseql (perhaps in another tickets)
(Utf8, Date)

I agree -- I can do so.

File: datafusion/expr/src/type_coercion/binary.rs (hunk 4 above)
Comment (comphead) [reply datafusion/expr/src/type_coercion/binary.rs:561]: We need another cast to add Utf <-> Timestamp #5117 probably addresses this cast. If not I'll create another ticket

File: datafusion/expr/src/type_coercion/binary.rs (hunk 5 above)
Comment (comphead) [reply datafusion/expr/src/type_coercion/binary.rs:567]: Thats correct, we lose TZ for now, we need some extra efforts to make arrow-rs work with TimestampTZ,planning to do so just after this PR merged.

File: datafusion/core/tests/sql/timestamp.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/tests/sql/timestamp.rs:1745]: Thank you very much -- that would be great

**Comment by alamb** [comment]
> We need another cast to add Utf <-> Timestamp https://github.com/apache/arrow-datafusion/pull/5117 probably addresses this cast. If not I'll create another ticket

I am going to merge this PR and test using https://github.com/apache/arrow-datafusion/pull/5117 and if it doesn't work I'll file another ticket

File: datafusion/expr/src/type_coercion/binary.rs (hunk 4 above)
Comment (alamb) [reply datafusion/expr/src/type_coercion/binary.rs:561]: Even after https://github.com/apache/arrow-datafusion/pull/5117 this still fails. I will file a follow on ticket

❯ select '2000-01-01T00:00:00'::timestamp::timestamptz = '2000-01-01T00:00:00';
Internal("The type of Timestamp(Nanosecond, Some(\"+00:00\")) Eq Utf8 of binary physical should be same")

Comment (alamb) [reply datafusion/expr/src/type_coercion/binary.rs:561]: Filed https://github.com/apache/arrow-datafusion/issues/5164

**Comment by alamb** [comment]
Filed https://github.com/apache/arrow-datafusion/issues/5164

File: datafusion/core/tests/sql/timestamp.rs (hunk 1 above)
Comment (comphead) [reply datafusion/core/tests/sql/timestamp.rs:1745]: Filed https://github.com/apache/arrow-datafusion/issues/5165

**Pull request merged**